stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,753 @@
+                // being busy. This scenario is extremely unlikely, because in order for it to happen, the application
+                // would need to transmit 4 (four) or more CAN frames with different CAN ID ordered from high ID to
+                // low ID nearly at the same time. For example:
+                //  1. 0x123        <-- Takes mailbox 0 (or any other)
+                //  2. 0x122        <-- Takes mailbox 2 (or any other)
+                //  3. 0x121        <-- Takes mailbox 1 (or any other)
+                //  4. 0x120        <-- INNER PRIORITY INVERSION HERE (only if all three TX mailboxes are still busy)
+                // This situation is even less likely to cause any noticeable disruptions on the CAN bus. Despite that,
+                // it is better to warn the developer about that during debugging, so we fire an assertion failure here.
+                // It is perfectly safe to remove it.
+                // Note: this error condition (all TX mailboxes busy) also occurs when erroneously sending data to the
+                // TX mailboxes of a non-existing CAN interface. Eg: accessing CAN2 on a device that only has CAN1.
+#if BXCAN_CONFIG_DEBUG_INNER_PRIORITY_INVERSION
+                BXCAN_ASSERT(!"CAN PRIO INV");  // NOLINT
+#endif
+                prio_inversion = true;
+            }
+            else
+            {
+                prio_inversion = false;
+            }
+        }
+        else  // All TX mailboxes are free, use first
+        {
+            tx_mailbox = 0U;
+        }
+
+        BXCAN_ASSERT(tx_mailbox < 3U);  // Index check - the value must be correct here
+    }
+
+    // By this time we've proved that a priority inversion would not occur, and we've also
+    // found a free TX mailbox. Therefore it is safe to enqueue the frame now.
+    if ((input_ok == true) && (prio_inversion == false))
+    {
+        volatile BxCANTxMailboxType* const mb = &bxcan_base->TxMailbox[tx_mailbox];
+
+        // The payload can be any length up to the MTU. We first make a local copy of the
+        // available payload bytes in a zero-filled array of MTU size. This makes the logic
+        // for filling the mailbox simpler as the payload pointer can be to variable-size
+        // data or even be a NULL pointer.
+        static uint8_t scratch_data[8] = {0};  // Not strictly needed to zero the storage.
+        if (payload_size > 0U)                 // The check is needed to avoid calling memcpy() with a NULL pointer.
+        {
+            // Clang-Tidy raises an error recommending the use of memcpy_s() instead.
+            // We ignore it because the safe functions are poorly supported; reliance on them may limit the portability.
+            (void) memcpy(&scratch_data[0], payload, payload_size);  // NOLINT
+        }
+
+        mb->TDTR = payload_size;  // DLC equals data length except in CAN FD
```

Comment:
TDTR is a time register, not length.

Reply:
Indeed, but the 3 LSB are the DLC. Thus, I believe that line to be correct.

However, reviewing the documentation, there is one niggle: bits 15:9 and 7:4 must be kept at reset value, and that value is not listed in any of the register tables (it says "-"). Also, the fields are not listed as writeable.  Thus, effectively writing zeros into these, as is happening now, is not right and probably undefined.

I propose the following:
// TDTR contains the DLC bits. However, there are reserved bits in the register that should not be changed.
// Selectively clear and set the DLC bits.
mb->TDTR &= ~BXCAN_TDTR_DLC_MASK;
mb->TDTR |= (payload_size & BXCAN_TDTR_DLC_MASK) < BXCAN_TDTR_DLC_SHIFT; // DLC equals data length except in CAN FD.

From the ST documentation:
CAN mailbox data length control and time stamp register (CAN_TDTxR) (x = 0..2)
Bits 31:16 TIME[15:0]: Message time stamp. Timer value captured at the SOF transmission.
Bits 15:9 Reserved, must be kept at reset value.
Bit 8 TGT : Transmit global time
Bits 7:4 Reserved, must be kept at reset value.
Bits 3:0 DLC[3:0]: Data length code

Reply:
>Indeed, but the 3 LSB are the DLC. Thus, I believe that line to be correct.

True. Sorry, I forgot that it is a multi-purpose register.

> However, reviewing the documentation, there is one niggle: bits 15:9 and 7:4 must be kept at reset value, and that value is not listed in any of the register tables (it says "-"). Also, the fields are not listed as writeable (blank). Thus, effectively writing zeros into these, as is happening now, is not right and probably undefined.

This is very true. I still very much remember a really bad case from many years ago when I spent two days debugging an RS485 driver on some LPC chip that was misbehaving terribly because I neglected a similar line in the reference manual. :scream_cat: 

The v0 driver is a poor reference in this regard.

What you wrote maybe a little verbose but it seems functionally correct. If not sure, use this:

```c
mb->TDTR = (mb->TDTR & ~BXCAN_TDTR_DLC_MASK) | payload_size;
```

Reply:
Going through the code, there were many instances of the reserved bits not being respected. I've modified all of them, cross-referencing the register maps for my MCU (single CAN) and another one (dual CAN). Due to the differences in number of filters, I had to make the checks depend on the configured maximum interface index.

Yes, I can vividly imagine the misery debugging that sort of thing. (Once, I had the pleasure of designing my own PCI card. Which meant that the hardware (FPGA-based) + HDL, linux kernel driver, and user-space code was all my problem. Debugging that was special... But, at least I had detailed knowledge of what every level -should- be doing.)